numba/core/inline_closurecall.py

```diff
@@ -407,11 +407,17 @@ def copy_ir(the_ir):
         assert(len(callee_scopes) == 1)
         callee_scope = callee_scopes[0]
         var_dict = {}
-        for var in callee_scope.localvars._con.values():
+        for var in tuple(callee_scope.localvars._con.values()):
             if not (var.name in callee_freevars):
                 inlined_name = _created_inlined_var_name(
                     callee_ir.func_id.unique_name, var.name)
+                # Update the caller scope with the new names
                 new_var = scope.redefine(inlined_name, loc=var.loc)
+                # Also update the callee scope with the new names. Should the
+                # type and call maps need updating (which requires SSA form) the
+                # transformation to SSA is valid as the IR object is internally
+                # consistent.
+                callee_scope.redefine(inlined_name, loc=var.loc)
                 var_dict[var.name] = new_var
         self.debug_print("var_dict = ", var_dict)
         replace_vars(callee_blocks, var_dict)
```

Comment:
The same effect can be achieved by:

```suggestion
                # Also update the callee scope with the new names. Should the
                # type and call maps need updating (which requires SSA form) the
                # transformation to SSA is valid as the IR object is internally
                # consistent.
                # callee_scope.redefine(inlined_name, loc=var.loc)
                var_dict[var.name] = new_var
        self.debug_print("var_dict = ", var_dict)
        replace_vars(callee_blocks, var_dict)

        from numba.core.ir_utils import fixup_var_define_in_scope
        fixup_var_define_in_scope(callee_blocks)
```

Perhaps, `fixup_var_define_in_scope` should be part of `replace_vars`